### ISSUE TITLE
12.5-style custom resources - resources may be declared many times but only belong to a single scope

### ISSUE THREAD

**Issue by Poohblah** [chef/chef#4718]
## Description

12.5-style fused resources fail when a service definition is defined multiple times in different scopes.

Let's say that I need to write a recipe that creates a firewall rule on port 5678 and consumes a resource - call it 'foo' - which is defined in an upstream cookbook. I'm treating the resource as a black box. To create a firewall rule, I need to include the default firewall recipe and execute the `firewall_rule` resource.  However, unbeknownst to me, the foo resource *also* needs to create a firewall rule (this firewall rule happens to be on port 1234), so the foo resource *also* includes the default firewall recipe.  This is where the conflict occurs. The firewall recipe only gets included once; however, since my cookbook includes the recipe before the foo resource, then the service is defined in the `run_context` of my recipe and not the `run_context` of the foo resource.  This causes the foo resource to be unable to find the firewall service and converge to fail.

I believe this may be related to https://github.com/chef/chef/issues/4669.

## Chef Version

12.8.1 via chef-dk 0.11.2

## Platform Version

Only tested with Linux chef-dk using Linux guest vms run via testkitchen for convergence

## Replication Case

`cookbooks/test/resources/lwrp.rb`:

```
action :foo do
  include_recipe 'firewall'

  firewall_rule 'foo' do
    port 1234
  end
end
```

`cookbooks/consumes_test/recipes/default.rb`:

```
include_recipe 'firewall'

firewall_rule 'bar' do
  port 5678
end

test_lwrp 'baz'
```

## Client Output

```
[birdsnest /tmp/consumes_test](master|…)[I]% kitchen converge       
-----> Starting Kitchen (v1.5.0)
-----> Converging <default-centos-71>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 4.2.1...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-centos-71>
       Starting Chef Client, version 12.8.1
       resolving cookbooks for run list: ["consumes_test::default"]
       Synchronizing Cookbooks:
         - consumes_test (0.1.0)
         - firewall (2.5.0)
         - test (0.1.0)
         - chef-sugar (3.3.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Recipe: chef-sugar::default
         * chef_gem[chef-sugar] action install (up to date)
         Converging 9 resources
         * chef_gem[chef-sugar] action install (up to date)
       Recipe: firewall::default
         * firewall[default] action install
           - install firewalld, create template for /etc/sysconfig
         * yum_package[firewalld] action install (up to date)
         * service[firewalld] action enable (up to date)
         * service[firewalld] action start
           - start service service[firewalld]
         * file[create empty /etc/sysconfig/firewalld-chef.rules] action create (skipped due to not_if)
         * firewall_rule[allow world to ssh] action create
         
         * firewall_rule[allow world to winrm] action create (skipped due to only_if)
         * firewall_rule[allow world to mosh] action create (skipped due to only_if)
         * firewall_rule[established] action create
         
         * firewall_rule[ipv6_icmp] action create
         
       Recipe: consumes_test::default
         * firewall_rule[bar] action create
         
         * test_lwrp[baz] action foo
           * firewall_rule[foo] action create
             
             ================================================================================
             Error executing action `create` on resource 'firewall_rule[foo]'
             ================================================================================
             
             Chef::Exceptions::ResourceNotFound
             ----------------------------------
             Cannot find a resource matching firewall[default] (did you define it first?)
             
             Cookbook Trace:
             ---------------
             /tmp/kitchen/cache/cookbooks/firewall/libraries/provider_firewall_rule.rb:27:in `block in <class:FirewallRuleGeneric>'
             
             Resource Declaration:
             ---------------------
             # In /tmp/kitchen/cache/cookbooks/test/resources/lwrp.rb
             
        4:   firewall_rule 'foo' do
        5:     port 1234
        6:   end
        7: end
             
             Compiled Resource:
             ------------------
             # Declared in /tmp/kitchen/cache/cookbooks/test/resources/lwrp.rb:4:in `block in class_from_file'
             
             firewall_rule("foo") do
        action [:create]
        retries 0
        retry_delay 2
        default_guard_interpreter :default
        declared_type :firewall_rule
        cookbook_name "consumes_test"
        port 1234
        notify_firewall true
        firewall_name "default"
             end
             
           
           ================================================================================
           Error executing action `foo` on resource 'test_lwrp[baz]'
           ================================================================================
           
           Chef::Exceptions::ResourceNotFound
           ----------------------------------
           firewall_rule[foo] (/tmp/kitchen/cache/cookbooks/test/resources/lwrp.rb line 4) had an error: Chef::Exceptions::ResourceNotFound: Cannot find a resource matching firewall[default] (did you define it first?)
           
           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/firewall/libraries/provider_firewall_rule.rb:27:in `block in <class:FirewallRuleGeneric>'
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/consumes_test/recipes/default.rb
           
            13: test_lwrp 'baz'
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/consumes_test/recipes/default.rb:13:in `from_file'
           
           test_lwrp("baz") do
             action [:foo]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :test_lwrp
             cookbook_name "consumes_test"
             recipe_name "default"
           end
           
       Recipe: firewall::default
         * firewall[default] action restart
            (skipped due to only_if)
          (skipped due to only_if)
       * file[/etc/sysconfig/firewalld-chef.rules] action create
         - update content in file /etc/sysconfig/firewalld-chef.rules from 946196 to 1c2b57
         --- /etc/sysconfig/firewalld-chef.rules	2016-03-16 17:11:16.331129951 +0000
         +++ /etc/sysconfig/.chef-firewalld-chef.rules20160316-14566-rw5ffp	2016-03-16 17:13:16.657262951 +0000
         @@ -1,4 +1,5 @@
          # position 50
         +firewall-cmd --direct --add-rule ipv4 filter INPUT 50 -p tcp -m tcp -m multiport --dports 22 -m comment --comment 'allow world to ssh' -j ACCEPT
          firewall-cmd --direct --add-rule ipv4 filter INPUT 50 -m state --state RELATED,ESTABLISHED -m comment --comment 'established' -j ACCEPT
          firewall-cmd --direct --add-rule ipv6 filter INPUT 50 -m state --state RELATED,ESTABLISHED -m comment --comment 'established' -j ACCEPT
          firewall-cmd --direct --add-rule ipv6 filter INPUT 50 -p ipv6-icmp -m comment --comment 'ipv6_icmp' -j ACCEPT
         - restore selinux security context
       [2016-03-16T17:13:16+00:00] WARN: Cloning resource attributes for service[firewalld] from prior resource (CHEF-3694)
       [2016-03-16T17:13:16+00:00] WARN: Previous service[firewalld]: /tmp/kitchen/cache/cookbooks/firewall/libraries/provider_firewall_firewalld.rb:38:in `block (2 levels) in <class:FirewallFirewalld>'
       [2016-03-16T17:13:16+00:00] WARN: Current  service[firewalld]: /tmp/kitchen/cache/cookbooks/firewall/libraries/provider_firewall_firewalld.rb:92:in `block in <class:FirewallFirewalld>'
       
       
       Running handlers:
       [2016-03-16T17:13:19+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2016-03-16T17:13:19+00:00] ERROR: Exception handlers complete
       Chef Client failed. 8 resources updated in 06 seconds
       [2016-03-16T17:13:19+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2016-03-16T17:13:19+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2016-03-16T17:13:19+00:00] ERROR: test_lwrp[baz] (consumes_test::default line 13) had an error: Chef::Exceptions::ResourceNotFound: firewall_rule[foo] (/tmp/kitchen/cache/cookbooks/test/resources/lwrp.rb line 4) had an error: Chef::Exceptions::ResourceNotFound: Cannot find a resource matching firewall[default] (did you define it first?)
       [2016-03-16T17:13:21+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
>>>>>> Converge failed on instance <default-centos-71>.
>>>>>> Please see .kitchen/logs/default-centos-71.log for more details
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: SSH exited (1) for command: [sh -c '

sudo -E /opt/chef/bin/chef-client --local-mode --config /tmp/kitchen/client.rb --log_level warn --force-formatter --no-color --json-attributes /tmp/kitchen/dna.json --chef-zero-port 8889
']
>>>>>> ----------------------
zlib(finalizer): the stream was freed prematurely.
```

## Stacktrace

```
[vagrant@default-centos-71 ~]$ sudo cat /tmp/kitchen/cache/chef-stacktrace.out
Generated at 2016-03-16 17:13:19 +0000
Chef::Exceptions::ResourceNotFound: test_lwrp[baz] (consumes_test::default line 13) had an error: Chef::Exceptions::ResourceNotFound: firewall_rule[foo] (/tmp/kitchen/cache/cookbooks/test/resources/lwrp.rb line 4) had an error: Chef::Exceptions::ResourceNotFound: Cannot find a resource matching firewall[default] (did you define it first?)
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:63:in `lookup'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:142:in `block (2 levels) in find_resource_by_hash'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:141:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:141:in `block in find_resource_by_hash'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:139:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:139:in `find_resource_by_hash'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:85:in `block in find'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:82:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_set.rb:82:in `find'
/tmp/kitchen/cache/cookbooks/firewall/libraries/provider_firewall_rule.rb:27:in `block in <class:FirewallRuleGeneric>'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/lwrp_base.rb:87:in `instance_eval'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider/lwrp_base.rb:87:in `block in action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider.rb:144:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource.rb:596:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:73:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:84:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:82:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:104:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider.rb:361:in `compile_and_converge_action'
(eval):2:in `action_foo'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/provider.rb:144:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource.rb:596:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:73:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:105:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:84:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/resource_collection/resource_list.rb:82:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/runner.rb:104:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/client.rb:668:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/client.rb:663:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/client.rb:663:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/client.rb:702:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/client.rb:280:in `run'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:270:in `block in fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:258:in `fork'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:258:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:223:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:211:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/client.rb:445:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/client.rb:435:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/client.rb:435:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application/client.rb:424:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/lib/chef/application.rb:58:in `run'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.8.1/bin/chef-client:26:in `<top (required)>'
/opt/chef/bin/chef-client:50:in `load'
/opt/chef/bin/chef-client:50:in `<main>'
```

**Comment by lamont-granquist** [comment]
can you replicate this with old style LWRPs (where the provider declares use_inline_resources)? 

because it smells to me like the cookbook was written so that you can't put firewall rules into providers that have sub-resource collections since the firewall service only exists in the outer run context.

**Comment by Poohblah** [comment]
Before I answer your question, I should point out that the firewall cookbook (which is an official Chef cookbook) is not the only upstream cookbook that's triggered this issue.  I've had the same experience with the apache2 cookbook, for instance.

To answer your question:
* I **do** run into the issue using an old-style LWRP **with** `use_inline_resources`.
* I **do not** run into the issue using an old-style LWRP **without** `use_inline_resources`.
* I **do not** run into the issue using a definition instead of an LWRP.

**Comment by lamont-granquist** [comment]
So this is basically a dup of #4017 then, and nothing new with 12.5-style resources.

**Comment by coderanger** [comment]
I think the core problem here is `I need to include the default firewall recipe`. That's not really supported in a bunch of different ways.

**Comment by Poohblah** [comment]
@coderanger you mean from inside an LWRP?

**Comment by coderanger** [comment]
Yes.

**Comment by coderanger** [comment]
Poise has a helper for it that works a bit better at least, but it's still not perfect. Notably I think the `firewall` cookbook was originally written using my helpers and then ported back to just core Chef so some bits might have been lost in translation.

**Comment by Poohblah** [comment]
So if my resource has a requirement to include another recipe, should I be using something other than a resource + provider? The only things that spring to mind are definitions and recipes, but recipes can't easily be parameterized and definitions are kinda-sorta deprecated.

**Comment by coderanger** [comment]
I mean my solution would be (and is) to use my `include_recipe` helpers for this, but just do so very sparingly. This would be a situation that might justify it, but it depends on what you want the UX to be like if no parent resource can be found (error vs. auto-populate). https://github.com/poise/poise-monit/blob/master/lib/poise_monit/service_providers/monit.rb#L56-L66 is an example of this kind of pattern.

**Comment by Poohblah** [comment]
Before I filed this bug report, when writing my resource, I did something quite similar to that helper as a temporary workaround to help alleviate the issues that consumers of my resource were running into. I'll see if I can't leverage Poise instead of using my own custom solution.

**Comment by lamont-granquist** [comment]
So I've got fixes in flight for most of this problem.  But there's still going to be a race condition.

We compile all the resources in the root run_context.  In this case the default firewall recipe comes before the test_lwrp call.  By the time we are in the provider for test_lwrp the default firewall recipe has already converged.

If this is an accumulator pattern, then the way to fix this is probably to have the default firewall recipe do all its work via a delayed notification so that it runs very late in the root run_context and then the stuff that is in flight will let you edit the resource in the root run context to accumulate config into it.

That's the only pattern I can come up with if some resource needs to really run ONE time and needs to accumulate state from multiple LWRPs.

If you it would be acceptable to run the firewall recipe multiple times (and possibly just let idempotence checks take care of the problem) then the problem is that you shouldn't be using a recipe.  That is just fighting the check which ensures that a given recipe runs only once.  You can convert the recipe to a resource and call that instead of calling include_recipe from within providers, or put the resource into a ruby mixin/library and include that call where you need to (which boils down to just rolling your own definition in pure ruby, but not mixing it into the entire DSL and bypassing the awful params mangling that definitions do, and avoiding use_inline_resources concerns and including the resources directly into the run_context that you mix the library into).  Either way if you use a resource, a definition or a method you're fundamentally avoiding the i-can-only-be-run-once check that recipes do -- which is the problem with writing reusable library code as recipes, and why it should probably never be done.  The default recipe in the firewall cookbook is likely a design mistake one way or the other.

**Comment by Poohblah** [comment]
Not sure I understand you here.  I don't believe the reported issue has anything to do with race conditions or the design of the upstream cookbook that's being included here. I understand the issue to be entirely a scoping issue, or more specifically, that including a recipe only adds the resources to the scope where the recipe is first included, even if a recipe is included in multiple scopes, due to the way the including recipes works, or the "i-can-only-be-run-once check" as you call it.

**Comment by lamont-granquist** [comment]
right but the "i-can-only-be-run-once-check" isn't ever going to be changed.  when you run into that, the solution is to rewrite the recipe as a resource -- using a  recipe is picking the wrong abstraction.

**Comment by Poohblah** [comment]
> the solution is to rewrite the recipe as a resource

That's one of the things I don't understand about your previous comment.  In this case, I'm not writing a recipe, I'm writing a resource and consuming an upstream recipe that I didn't write (the default recipe of the firewall cookbook).  I understand why the recipe might be the wrong abstraction but when consuming an upstream cookbook there's nothing I can do about it.

**Comment by Poohblah** [comment]
Now we've circled back to one of my previous comments:

> So if my resource has a requirement to include another recipe, should I be using something other than a resource + provider? The only things that spring to mind are definitions and recipes, but recipes can't easily be parameterized and definitions are kinda-sorta deprecated.

**Comment by lamont-granquist** [comment]
re: `when consuming an upstream cookbook there's nothing I can do about it.`

submit a PR to the cookbook and fix it.  extract the recipe into a resource or definition/method and replace the default recipe with a call to that.

e.g. https://github.com/chef-cookbooks/aws/pull/231

**Comment by Poohblah** [comment]
I don't think that's a tenable solution.  I've run into this issue when consuming many different community cookbooks, and forking all of them is going to be a tracking & maintenance nightmare.  For instance, the apache2 cookbook, [which exposes almost all of its consumables as recipes][apache2], and those recipes are non-trivial to port to resources.  Not to mention that I seriously doubt that such PR is going to be accepted, since such a large change would break every single cookbook that depends on the apache2 cookbook.

Maybe I've got my head in the clouds here, but I really do think that "I can't change upstream or how I consume upstream" should be a **very** strong assumption for the sake of this particular issue.

Or maybe I'm misunderstanding you. Perhaps you're suggesting a temporary workaround and not a permanent solution?

[apache2]: https://github.com/svanzoest-cookbooks/apache2/tree/master/recipes

**Comment by Poohblah** [comment]
I'm not too savvy on on the details of Chef internals, but it looks like https://github.com/chef/chef/pull/4741 would fix this issue?

**Comment by lamont-granquist** [comment]
I didn't say fork the cookbook.  I said submit a PR to fix the upstream cookbook.  They're community cookbooks and you're part of the community, so you own them, and can maintain them.  They have lots of code which is, honestly, completely shitty and relies on engineering that was done 6+ years ago when nobody knew any better way of doing things.   The PR to the AWS cookbook to remove the need for the deafult recipe is a good example of how to go about this.

The apache cookbook, in particular, is a disaster for this very reason.  That is why the httpd cookbook exists and was written:  https://github.com/chef-cookbooks/httpd

And I don't have a solution and I don't think anyone has a solution that can salvage the apache2 cookbook -- that's why we wrote a replacement.

**Comment by lamont-granquist** [comment]
#4741 gets us in a better state, but it won't fix all the problems.

**Comment by lamont-granquist** [comment]
Think i'm going to close this, #4741 / #4017 have been merged.  I'm going to work on some other API extensions to how we manipulate resources to make writing accumulator-like-thingies and manipulating your parent or the root run_context's resource_collections a bit easier along with maybe slurping in some more poise-like features.  I'm not sure those latter things are directly related to this, but might make it easier.  I also can't extract anything out of this that's actionable right now.

I have a TODO to blog some about this stuff once I get done tweaking things, and then hopefully people will play around with stuff and we'll find some patterns to try and issues with them and I'm sure it'll come back around again.  The firewall cookbook rule is something that I'm probably going to look at fixing at some point in the future, as I'm starting to use it in my own cookbooks and I'm certain that with some familiarity will come some code rage to fix it all..

**Comment by Poohblah** [comment]
Thanks, @lamont-granquist. And thanks for being patient and answering all my questions.  I still disagree with some of the things you've said regarding submitting PRs to upstream as solutions to specific problems, but it looks to me like #4741 should help quite a bit.

**Comment by lamont-granquist** [comment]
I think we're going to have to do that.  At some point I think the community will need to realize that reusable library cookbooks that have recipes and/or attributes directories in them are simply poorly written.  The problem is that about the only people writing resource-driven cookbooks are Chef employees and we don't even have the (human) resources to update all of the chef-cookbooks.  I don't know how to solve the problem of taking a cookbook written with recipes and attributes and fixing it though.  I would clearly prefer to do that rather than just yell at people that their cookbooks are all bad...
